### DISCUSSION TITLE
API Search Repositories with no duplicate repos. it's possible?

### DISCUSSION THREAD

**Discussion by StephLang99** [community/community#24361] [Apps, API and Webhooks]
<p>Hi,<br>
I doing a search on github api repositories with my python program.<br>
the result from this search is limited to 1000. this is correct.<br>
but in this result of 1000 row. the same repositories (same URL) is  there.<br>
it’s possible to have a search with 1000 repos with no duplicate?</p>
<p>my query : Query = language:python<br>
here a sample.<br>
"…<br>
“https:///github.com/pennersr/django-allauth.git”<br>
“https:///github.com/jazzband/django-debug-toolbar.git”<br>
“https:///github.com/jazzband/django-debug-toolbar.git &lt;- duplicate”<br>
“https:///github.com/jazzband/django-debug-toolbar.git &lt;- duplicate”<br>
“https:///github.com/django-extensions/django-extensions.git”<br>
“https:///github.com/django-oscar/django-oscar.git”<br>
“https:///github.com/jupyter/docker-stacks.git”<br>
“https:///github.com/jupyter/docker-stacks.git &lt;- duplicate”<br>
“https:///github.com/docopt/docopt.git”<br>
“https:///github.com/shidenggui/easytrader.git”<br>
“https:///github.com/edgedb/edgedb.git”<br>
“https:///github.com/edgedb/edgedb.git  &lt;- duplicate”</p>

**Comment by lee-dohm** [comment]
<p>Hello <a class="mention" href="/stephlang99">@StephLang99</a> and welcome to the community.</p>
<p>Can you give the exact URL you’re making the request to in order to receive these results? That will help us debug what’s going on here better.</p>
<p>Let us know!</p>

**Comment by StephLang99** [comment]
<p>Hi thanks to help me.<br>
here <a href="https://api.github.com/search/repositories?q=language:python&amp;sort=stars&amp;order=desc&amp;page=10&amp;per_page=100" rel="noopener nofollow ugc">https://api.github.com/search/repositories?q=language:python&amp;sort=stars&amp;order=desc&amp;page=10&amp;per_page=100</a></p>

**Comment by lee-dohm** [comment, marked as answer]
<p>Thanks for sharing the URL. I tried the following and, at least in the first 100 results, there are no duplicate repositories:</p>
<pre><code class="lang-bash">$ http -pb "https://api.github.com/search/repositories?q=language:python&amp;sort=stars&amp;order=desc&amp;per_page=100" | jq .items[].html_url | wc -l
     100
$ http -pb "https://api.github.com/search/repositories?q=language:python&amp;sort=stars&amp;order=desc&amp;per_page=100" | jq .items[].html_url | uniq | wc -l
     100
</code></pre>
<p>Are you certain you’re collating or parsing the results correctly? It might also be that, depending on how you paginate, that the sorted results could shift between requests and that might be the cause of the duplicates.</p>

**Comment by StephLang99** [comment]
<p>yes, you’re right.<br>
the issue is with the library pygithub when is handling the paginate.<br>
I displayed the stars and the counter is not  descending.</p>

**Comment by di-press** [comment]
When I try to query for 1000 repositories, sorted by the descending number of stars and by 'jupyter-notebook' language, I don't get unique results. I only get 931 unique results, and the remaining is repeated. 

``` def retrieve_top_1000_repositories(filepath: str):

    # GitHub API endpoint to get the top 1000 starred repos with 'jupyter-notebook' language
    url = 'https://api.github.com/search/repositories?q=language:jupyter-notebook&sort=stars&order=desc&per_page=100'
    headers = {"Authorization": AUTH_TOKEN}

    # Send the first GET request to the API endpoint
    response = requests.get(url, headers=headers)

    # Check if the request was successful (status code 200)
    if response.status_code == 200:
        # Get the JSON data from the response
        data = response.json()

        with open('top_1000_repositories.json', 'w') as f:
            json.dump(response.json(), f)
            print('Repository data saved to top_1000_repositories.json')
        
        # Extract the URL of each repository and save it in a list
        urls = [repo['html_url'] for repo in data['items']]
        
        # Redirecting to the next page of results
        # Retrieve the remaining repositories and add their URLs to the list.
        # SInce the results are repeated, 
        for i in range(1, 11):
            # Send a GET request to the next page of results
            response = requests.get(url + '&page=' + str(i+1), headers=headers)
            
            # Check if the request was successful (status code 200)
            if response.status_code == 200:
                # Get the JSON data from the response
                data = response.json()

                with open(("top_1000_repositories" + str(i)+ '.json'), 'a+') as f:
                    json.dump(data, f)
                    print('Repository data saved to top_1000_repositories.json')
                    
                # Extract the URL of each repository and add it to the list
                urls.extend([repo['html_url'] for repo in data['items']])
            else:
                print('Error:', response.status_code)
        
        # Save the URLs in a txt file
        with open(filepath, 'w') as f:
            index = 1
            for url in urls:
                f.write(url + " " + str(index) + '\n')
                index += 1
                
        print(f'Results saved in {filepath}')
    else:
        print('Error:', response.status_code)
    return urls
```

As the GitHub API limits the search up to 1000 results, I receive a 403 status code when accessing the last page of results (11th page), but that is expected. What I didn't expect was the repeated results. I couldn't find a way to avoid getting duplicated results.
I've tried this same task in shell code, using the 'uniq' parameter to avoid repeated results, but it wasn't successful.

When I request only 100 results, they are unique.

**Comment by navrkald** [comment]
Hi everyone and @di-press , I have reported this bug to Github and they indeed confirmed there is the bug in Github Search Rest API when fetching multiple pages of results, however they haven't provided any timeline to fix. Looks to me like very old issue which nobody from Github side was able to fix yet....
